src/codegen/generators/expression_generator.rs

```diff
@@ -745,6 +750,58 @@ impl<'ink, 'b> ExpressionCodeGenerator<'ink, 'b> {
         .map(Into::into)
     }
 
+    pub fn generate_variadic_arguments_list(
+        &self,
+        pou: &PouIndexEntry,
+        variadic_params: &[&AstStatement],
+    ) -> Result<Vec<BasicValueEnum<'ink>>, Diagnostic> {
+        let generated_params = variadic_params
+            .iter()
+            .map(|param_statement| {
+                self.get_type_hint_for(param_statement)
+                    .map(|it| it.get_name())
+                    .and_then(|type_name| self.generate_argument_by_val(type_name, param_statement))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        //get the real varargs from the index
+        if let Some(VarArgs::Sized(Some(type_name))) = self
+            .index
+            .get_variadic_member(pou.get_name())
+            .and_then(|it| it.get_varargs())
+        {
+            let llvm_type = if pou.is_generic() {
+                //Use the type of the first parameter
+                let type_info = self.get_type_hint_info_for(variadic_params[0])?;
+                self.llvm_index.get_associated_type(type_info.get_name())
+            } else {
+                self.llvm_index.get_associated_type(type_name)
+            }?;
+            let size = generated_params.len();
+            let size_param = self
+                .llvm_index
+                .get_associated_type(typesystem::DINT_TYPE)
+                .expect("DINT always exists")
+                .into_int_type()
+                .const_int(size as u64, true);
+            let arr = Llvm::get_array_type(llvm_type, size as u32);
+            let arr_storage = self.llvm.builder.build_alloca(arr, "");
+            for (i, ele) in generated_params.iter().enumerate() {
+                let ele_ptr = self.llvm.load_array_element(
+                    arr_storage,
+                    &[
+                        self.llvm.context.i32_type().const_zero(),
+                        self.llvm.context.i32_type().const_int(i as u64, true),
+                    ],
+                    "",
+                )?;
+                self.llvm.builder.build_store(ele_ptr, *ele);
+            }
```

Comment:
llvm supports constant-arrays. I think runtime performance would be better

```llvm
@X = global i32 17
@Y = global i32 42
@Z = global [2 x i32*] [ i32* @X, i32* @Y ]
```

https://llvm.org/docs/LangRef.html#global-variable-and-function-addresses

Reply:
This does not work on non-constant variables (expressions that need to be resolved) it was the initial implementation and I had to scrape it

Reply:
why do you think so? did you try?
the inkwell type's are a bit opinionated. I just tried and at least I get the IR for the test `builtin_function_call_mux` generated:

```diff
   15    15 │   %load_c = load i32, i32* %c, align 4
   16    16 │   %load_d = load i32, i32* %d, align 4
   17    17 │   %load_e = load i32, i32* %e, align 4
   18    18 │   %1 = alloca [4 x i32], align 4
- 19        │-  %2 = getelementptr inbounds [4 x i32], [4 x i32]* %1, i32 0, i32 0
- 20        │-  store i32 %load_b, i32* %2, align 4
- 21        │-  %3 = getelementptr inbounds [4 x i32], [4 x i32]* %1, i32 0, i32 1
- 22        │-  store i32 %load_c, i32* %3, align 4
- 23        │-  %4 = getelementptr inbounds [4 x i32], [4 x i32]* %1, i32 0, i32 2
- 24        │-  store i32 %load_d, i32* %4, align 4
- 25        │-  %5 = getelementptr inbounds [4 x i32], [4 x i32]* %1, i32 0, i32 3
- 26        │-  store i32 %load_e, i32* %5, align 4
- 27        │-  %6 = getelementptr inbounds [4 x i32], [4 x i32]* %1, i32 0, i32 3
- 28        │-  %7 = load i32, i32* %6, align 4
- 29        │-  store i32 %7, i32* %a, align 4
+        19 │+  store [4 x i32] [i32 %load_b, i32 %load_c, i32 %load_d, i32 %load_e], [4 x i32]* %1, align 4
+        20 │+  %2 = getelementptr inbounds [4 x i32], [4 x i32]* %1, i32 0, i32 3
+        21 │+  %3 = load i32, i32* %2, align 4
+        22 │+  store i32 %3, i32* %a, align 4
   30    23 │   ret void
   31    24 │ }
```


It just works for int_values here, so i think you just need a helper method, where you pass the generated variadic parameters and it creates you an int_array or a ptr_array or a float_array etc. (depending on the variadic values you passed it --> see that all variadic parameters need to be the same type-nature anyway)

I think its a cleaner implementation and faster at runtime (and much easier for the optimizer)

here is my change:
```diff
index 9487bb518..893e72506 100644
--- a/src/codegen/generators/expression_generator.rs
+++ b/src/codegen/generators/expression_generator.rs
@@ -777,22 +777,19 @@ impl<'ink, 'b> ExpressionCodeGenerator<'ink, 'b> {
             } else {
                 self.llvm_index.get_associated_type(type_name)
             }?;
-            let size = generated_params.len();
-            let size_param = self.llvm.i32_type().const_int(size as u64, true);
-            let arr = Llvm::get_array_type(llvm_type, size as u32);
-            let arr_storage = self.llvm.builder.build_alloca(arr, "");
-            for (i, ele) in generated_params.iter().enumerate() {
-                let ele_ptr = self.llvm.load_array_element(
-                    arr_storage,
-                    &[
-                        self.llvm.context.i32_type().const_zero(),
-                        self.llvm.context.i32_type().const_int(i as u64, true),
-                    ],
-                    "",
-                )?;
-                self.llvm.builder.build_store(ele_ptr, *ele);
-            }
-            Ok(vec![size_param.into(), arr_storage.into()])
+            // generate the variadic values
+            let size_value = self.llvm.i32_type().const_int(generated_params.len() as u64, false);
+            let arr_storage = self.llvm.builder.build_array_alloca(llvm_type, size_value, "");
+            // these 2 lines need a helper function (it now assumes intValues)
+            let values: Vec<IntValue> = generated_params
+                .iter()
+                .map(|it| it.into_int_value())
+                .collect();
+            self.llvm.builder.build_store(arr_storage, llvm_type.into_int_type().const_array(values.as_slice()));
+
+            let size_param = self.llvm.builder.build_alloca(self.llvm.i32_type(), "");
+            self.llvm.builder.build_store(size_param, size_value);
+           Ok(vec![size_param.into(), arr_storage.into()])
         } else {
             Ok(generated_params)
         }
```

Reply:
There should be a correctness test for it. It will segfault or at least my implementation segfaults.
`lli`also refused to run it.
I didn't check your version yet maybe it's different I'll check tomorrow